esp-hal/src/gpio/mod.rs

```diff
@@ -883,22 +891,55 @@ impl InterruptConfigurable for Io {
 
 #[ram]
 extern "C" fn user_gpio_interrupt_handler() {
-    USER_INTERRUPT_HANDLER.call();
-
-    default_gpio_interrupt_handler();
+    handle_pin_interrupts(|| USER_INTERRUPT_HANDLER.call());
 }
 
 #[ram]
 extern "C" fn default_gpio_interrupt_handler() {
-    handle_pin_interrupts(on_pin_irq);
+    handle_pin_interrupts(|| ());
 }
 
 #[ram]
-fn on_pin_irq(pin_nr: u8) {
-    // FIXME: async handlers signal completion by disabling the interrupt, but this
-    // conflicts with user handlers.
-    set_int_enable(pin_nr, 0, 0, false);
```

Comment:
I'm very happy to see this go away from the interrupt handler but will this actually work for level events?

I may be wrong here but I don't think clearing the interrupt bit will be enough, as the interrupt bit would be immediately set again.

The HIL tests don't use the level events so I couldn't quickly prove myself wrong there.

Reply:
The fun with level events is, that they need to be disabled by the user. The point of them is that they keep firing over and over again while the pin's level is the one you listen for.

Reply:
Yeah, so `pin.wait_for_low().await;` or `pin.wait_for_high().await;` would never return in that case. (Except in cross-core scenarios)

(Have to say GPIO is my least favorite interrupt handler)

Reply:
Ah I get what you are asking - `pin.wait_for_high` will just end up looping. :(

Reply:
So, I guess I can be "hardware correct" here, use a level event and a separate flag to show that a pin needs to unlisten. Or, I can listen for an edge event, check if the pin is already the required level, and unlisten and resolve immediately if it is. For an async operation I don't think there would be an observable difference.

Reply:
I personally like the latter but yeah either would work I think.

The latter might mean less latency maybe? by skipping the `Poll::Pending`. Though there's other ways to accomplish that.